examples/genetic_music/sketch.js

```diff
@@ -0,0 +1,240 @@
+  // Sort in descending order of fitness
+  population.sort((a, b) => b.fitnessScore - a.fitnessScore);
+  // Keep only the N fittest
+  population = subset(population, 0, numberOfSurvivors);
+  // Re-assign ID numbers
+  for (var i=0; i<population.length; i++) {
+    population[i].id = i;
+  }
+}
+
+function reproducePopulation() {
+  var newPopulation = [];
+  while (newPopulation.length < maxPopulationSize - numberOfSurvivors) {
+    var parentA = random(population);
+    var parentB = random(population);
+    var child = parentA.reproduceWith(parentB);
+    newPopulation.push(child);
+  }
+  // Add new generation to the survivors
+  population = population.concat(newPopulation);
+  // Re-assign ID numbers
+  for (var i=0; i<population.length; i++) {
+    population[i].id = i;
+  }
+  generationCount++;
+}
+
+function fastForward() {
+  var fastForwardNum = 10;
+  for (var i=0; i<fastForwardNum; i++) {
+    selectFittest();
+    reproducePopulation();
+  }
+}
+
+function resetPopulation() {
+  generationCount = 1;
+  for (var i=0; i<maxPopulationSize; i++) {
+    var song = new Earworm(i);
+    song.initialize();
+    population[i] = song;
+  }
+}
+
+function Earworm(indexNumber) {
+  this.id = indexNumber;
+  this.length = songLength;
+  this.notes = [];
+  // Visual properties
+  this.xpos = random(width);
+  this.ypos = random(height);
+  this.radius = (width + height) / 50;
+  this.fitnessScore = 0;
+}
+Earworm.prototype.initialize = function() {
+  this.notes = [];
+  for (var i=0; i<this.length; i++) {
+    this.notes.push(random(validNotes));
+  }
+  this.calculateFitness();
+};
+Earworm.prototype.display = function() {
+  this.xpos = constrain(this.xpos + random(-1, 1), 0, width);
+  this.ypos = constrain(this.ypos + random(-1, 1), 0, height);
+
+  push();
+  strokeWeight(1);
+  angleMode(DEGREES); // Change the mode to DEGREES
+  var angle = 360 / this.notes.length;
+  translate(this.xpos, this.ypos);
+  for (var i=0; i<this.notes.length; i++) {
+    var color = map(this.notes[i], minValidNote, maxValidNote, 280, 120) % 255;
+    var length = map(this.notes[i], minValidNote, maxValidNote, this.radius/2, this.radius);
+    strokeWeight(1);
+    stroke(color, 180, 250);
+    if (songIsPlaying) {
+      if (this.id == clickedEarwormIndex) {
+        stroke(color, 180, 250);
+        if (i == notePlaybackIndex) {
+          strokeWeight(2);
+          length = this.radius;
+        } else {
+          strokeWeight(1);
+        }
+      } else {
+        stroke(color, 100, 100);
+      }
+    }
+    rotate(angle);
+    line(0, 0, length, 0);
+  }
+  pop();
+};
+Earworm.prototype.calculateFitness = function() {
+  this.fitnessScore = 0;
+  // Key
+  for (var i=0; i<this.notes.length; i++) {
+    var keyClass = this.notes[i] % 12;
+    if (desiredKeyClasses.indexOf(keyClass) >= 0) {
+      this.fitnessScore = this.fitnessScore + 10;
+    }
+  }
+  // Prefer smaller intervals
+  for (var i=0; i<this.notes.length-1; i++) {
```

Comment:
It seems like this has two goals:

1. Prefer melodies that are within a smaller range. Maybe 24 semitones (two octaves)
2. Within that range, prefer some intervals over others.

Smaller intervals are the most dissonant, so we might instead want to score based on the ratio of [consonant to dissonant intervals](https://www.howmusicreallyworks.com/Pages_Chapter_4/4_2.html#4.2.6).